interfaces/prompting/constraints.go

```diff
@@ -30,62 +31,90 @@ import (
 	"github.com/snapcore/snapd/strutil"
 )
 
-// Constraints hold information about the applicability of a rule to particular
-// paths or permissions. A request matches the constraints if the requested path
-// is matched by the path pattern (according to bash's globstar matching) and
-// the requested permissions are contained in the constraints' permissions.
+// Constraints hold information about the applicability of a new rule to
+// particular paths and permissions. When creating a new rule, snapd converts
+// Constraints to RuleConstraints.
 type Constraints struct {
-	PathPattern *patterns.PathPattern `json:"path-pattern,omitempty"`
-	Permissions []string              `json:"permissions,omitempty"`
+	PathPattern *patterns.PathPattern `json:"path-pattern"`
+	Permissions PermissionMap         `json:"permissions"`
```

Comment:
Yes. The `desktop-security-center` is the only supported client at the moment, and it will accept either the old or the new format and always map the result to the new format, since the new format can express a proper superset of what the old format can express. Because the feature is experimental and the API is not yet officially documented, we feel comfortable making this breaking change, which the only "official" client already handles correctly.